pkg/chaosdaemon/stress_server.go

```
@@ -0,0 +1,223 @@
+	log.Info("executing stressors", "request", req)
+	pid, err := s.crClient.GetPidFromContainerID(ctx, req.Target)
+	if err != nil {
+		return nil, err
+	}
+	path := PidPath(int(pid))
+	id, err := s.crClient.StripContainerIDProtocolPrefix(ctx, req.Target)
+	if err != nil {
+		return nil, err
+	}
+	cgroup, err := findValidCgroup(path, id)
+	if err != nil {
+		return nil, err
+	}
+	if req.Scope == pb.StressRequest_POD {
+		cgroup, _ = filepath.Split(cgroup)
+	}
+	control, err := cgroups.Load(cgroups.V1, cgroups.StaticPath(cgroup))
+	if err != nil {
+		return nil, err
+	}
+	cmd := exec.Command("stress-ng", strings.Fields(req.Stressors)...)
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	if err = control.Add(cgroups.Process{Pid: cmd.Process.Pid}); err != nil {
+		if err := cmd.Process.Kill(); err != nil {
+			return nil, err
+		}
+	}
+	stressorLocker.Lock()
+	defer stressorLocker.Unlock()
+	podStressors[req.Target] = cmd
+	go func() {
+		if err, ok := cmd.Wait().(*exec.ExitError); ok {
+			status := err.Sys().(syscall.WaitStatus)
+			if status.Signaled() && status.Signal() == syscall.SIGKILL {
+				log.Info("stressors cancelled", "request", req)
+			} else {
+				log.Error(err, "stressors exited accidentally", "request", req)
+			}
+		}
+		stressorLocker.Lock()
+		defer stressorLocker.Unlock()
+		delete(podStressors, req.Target)
+	}()
+
+	return &empty.Empty{}, nil
+}
+
+func (s *daemonServer) CancelStressors(ctx context.Context,
+	req *pb.StressRequest) (*empty.Empty, error) {
+	log.Info("canceling stressors", "request", req)
+	if cmd, ok := podStressors[req.Target]; ok {
+		if err := cmd.Process.Kill(); err != nil {
+			log.Error(err, "fail to exit stressors", "pid", cmd.Process.Pid)
+			return nil, err
+		}
+	}
+	return &empty.Empty{}, nil
+}
+
+func findValidCgroup(path cgroups.Path, target string) (string, error) {
+	for _, subsys := range cgroupSubsys {
+		if p, _ := path(cgroups.Name(subsys)); strings.Contains(p, target) {
```

Why ignore the error here?

---

I think we only need to find one cgroup path here. The `subsys` tried could not exist. So I think it's safe to ignore the error.

---

Got

---

I suggest not to ignore error here for any reason.

---

@Yisaer I don't get it. Could you explain this more specificially ?

---

The error may contains the situation which we haven't considered, it is dangerous to ignore error unless for a strong reason. I didn't see it here.

---

I think if we never find a match due to whatever happened, we finally give an error. From that point of view, I think that is ok.

---

Though we would finally get an error, but we would miss the root error which we ignored here. I don't think this is a good reason to ignore error here.

---

Almost LGTM. plz handle the error instead of ignoring the error.